nomic/data_operations.py

```diff
@@ -641,130 +574,199 @@ class AtlasMapTags:
     the associated pandas DataFrame.
     """
 
-    def __init__(self, projection: "AtlasProjection"):
+    def __init__(self, projection: "AtlasProjection", auto_cleanup: Optional[bool] = False):
         self.projection = projection
-        self.project = projection.project
-        self.id_field = self.projection.project.id_field
-        self._tb: pa.Table = projection._fetch_tiles().select([self.id_field])
+        self.dataset = projection.dataset
+        self.id_field = self.projection.dataset.id_field
+        # Pre-fetch tiles first upon initialization
+        self.projection._fetch_tiles(overwrite=False)
+        self.auto_cleanup = auto_cleanup
 
     @property
-    def df(self) -> pd.DataFrame:
-        """
+    def df(self, overwrite: Optional[bool]=False) -> pd.DataFrame:
+        '''
         Pandas DataFrame mapping each data point to its tags.
-        """
-
-        id_frame = self._tb.to_pandas()
-        tag_to_datums = self.get_tags()
-
-        # encoded contains a multi-hot vector withs 1 for all rows that contain that tag
-        encoded = {key: [] for key in list(tag_to_datums.keys())}
-        for id in id_frame[self.id_field]:
-            for key in encoded:
-                if id in tag_to_datums[key]:
-                    encoded[key].append(1)
+        '''
+        tags = self.get_tags()
+        tag_definition_ids = [tag["tag_definition_id"] for tag in tags]
+        if self.auto_cleanup:
+            self._remove_outdated_tag_files(tag_definition_ids)
+        for tag in tags:
+            self._download_tag(tag["tag_name"], overwrite=overwrite)
+        tbs = []
+        all_quads = list(self.projection._tiles_in_order(coords_only=True))
+        for quad in tqdm(all_quads):
+            quad_str = "/".join([str(q) for q in quad])
```

Comment:
Use pathlib instead of `"/".join` to create safe filesystem paths.

Reply:
yeppppp same mistake that i made last time 🤦. fixed